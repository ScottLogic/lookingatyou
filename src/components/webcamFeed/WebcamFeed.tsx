import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { IVideoState, SET_VIDEO, IDimensions, SET_DIMENSIONS } from '../../store/actions/video/types';
import { Dispatch } from 'redux';
import { getDimensions, getVideo } from '../../store/reducers/videoReducer';

interface IWebcamFeedProps {
  deviceId: string,
  onUserMedia: (stream: MediaStream) => void,
  onUserMediaError: (e: Error) => void,
  setDimensions: (dimensions: IDimensions) => void,
  setVideo: (video: HTMLVideoElement) => void,
  video: HTMLVideoElement | null,
  dimensions: IDimensions,
}

const mapStateToProps = (state: IVideoState) => {
  return {
    video: getVideo(state),
    dimensions: getDimensions(state),
  }
}

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    setDimensions: (dimensions: IDimensions) => dispatch({ type: SET_DIMENSIONS, payload: dimensions }),
    setVideo: (video: HTMLVideoElement) => dispatch({ type: SET_VIDEO, payload: video }),
  }
}

function WebcamFeed(props: IWebcamFeedProps) {
  const [video, setVideo] = useState();

  function getVideo(element: HTMLVideoElement) {
    if (element) {
      setVideo(video);
    }
  }

  async function getWebcam() {
    if (!video && !props.video) {
      try {
        const myStream = await getStream(props.deviceId);
        var streamSettings = myStream.getVideoTracks()[0].getSettings();
        if (streamSettings.height && streamSettings.width) {
          props.setDimensions({ width: streamSettings.width, height: streamSettings.height })
        }

  function getStream(deviceId: string) {
    return navigator.mediaDevices.getUserMedia({ video: { deviceId } })
  }

  useEffect(() => {
    getWebcam();
  })

  return (
    <video
      className={props.deviceId + ' hidden'}
      autoPlay={true}
      height={props.dimensions.height}
      width={props.dimensions.width}
      ref={getVideo}
    />
  )
}

export default connect(mapStateToProps, mapDispatchToProps)(WebcamFeed);
