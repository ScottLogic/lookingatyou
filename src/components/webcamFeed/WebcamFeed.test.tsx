import React from 'react';
import WebcamFeed from "./WebcamFeed";
import { mount, shallow } from "enzyme";
import jsdom from 'jsdom';

describe('WebcamFeed', () => {
  const streamSuccess = 'Stream received';
  const streamFailure = 'Unable to get stream';
  const deviceId = '123456';

  const mockOnUserMedia = jest.fn(() => streamSuccess);
  const mockOnUserMediaError = jest.fn().mockReturnValue(streamFailure);
  const mockMediaDevices = new jsdom.JSDOM().window.navigator.mediaDevices;
  const myMediaDevices = jest.fn().mockImplementation(() => {
  });


  const props = {
    deviceId: deviceId,
    onUserMedia: mockOnUserMedia,
    onUserMediaError: mockOnUserMediaError,
  }


  it('Should call onUserMediaError when no webcam detected', () => {
    mount(<WebcamFeed mediaDevices={mockMediaDevices} {...props} />);
    expect(mockOnUserMediaError.mock.results[0].value).toBe(streamFailure);
  });

  it('Should call onUserMedia when webcam detected', () => {
    let myMockMediaDevices = new jsdom.JSDOM().window.navigator.mediaDevices;
    mount(<WebcamFeed mediaDevices={myMockMediaDevices} {...props} />);
    expect(mockOnUserMedia.mock.results[0].value).toBe(streamSuccess);
  })

  it('Should render video object', () => {
    const wrapper = shallow(<WebcamFeed mediaDevices={mockMediaDevices} {...props} />);
    expect(wrapper).toMatchSnapshot();
  })
})
