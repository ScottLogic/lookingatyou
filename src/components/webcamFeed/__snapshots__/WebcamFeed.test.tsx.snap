// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebcamFeed Should render video object 1`] = `ShallowWrapper {}`;
